src/Microsoft.TemplateEngine.Utils/Json/JsonBuilder.cs

```diff
@@ -0,0 +1,66 @@
+using System;
+using System.Collections.Generic;
+using Microsoft.TemplateEngine.Abstractions.Json;
+
+namespace Microsoft.TemplateEngine.Utils.Json
+{
+    internal class JsonBuilder<TResult, TResultConcrete> : IJsonBuilder<TResult>
+        where TResultConcrete : TResult
+    {
+        private readonly Func<TResult> _creator;
+        private readonly Dictionary<string, Action<IJsonToken, TResult>> _deserializeSteps = new Dictionary<string, Action<IJsonToken, TResult>>(StringComparer.OrdinalIgnoreCase);
+        private readonly List<Action<TResult, IJsonObject>> _serializeSteps = new List<Action<TResult, IJsonObject>>();
+
+        public JsonBuilder(Func<TResult> creator)
+        {
+            _creator = creator;
+        }
+
+        public TResult Deserialize(IJsonToken source)
+        {
+            if (source.TokenType == JsonTokenType.Object)
+            {
+                TResult result = _creator();
+                ((IJsonObject)source).ExtractValues(result, _deserializeSteps);
+                return result;
+            }
+
+            return default(TResult);
+        }
+
+        public JsonBuilder<TResult, TResultConcrete> MapCore<T, TConcrete>(Getter<TResultConcrete, T> getter, Serialize<T> serialize, Deserialize<TConcrete> deserialize, Setter<TResultConcrete, T> setter, Func<TConcrete> itemCreator, string propertyName)
```

Comment:
Nit: Generic "T" could be more descriptive.

Reply:
Updated